### PR TITLE
Abort tests

### PR DESCRIPTION
**Pull request by Horneth**
Some refactoring and adding 3 abort tests (ignored for now):

1. Starts a workflow and aborts it immediately after the submission response, check that status is aborted
2. Starts a workflow, waits for the second call to be running, and abort the workflow.
Check that the status is aborted, wait a little and check metadata again to make sure there was no new calls started.
In JES land, check that JES says the operation was cancelled and that the VM is destroyed (with a 1 minute timeout)
3. Same as 2 but restart Cromwell immediately after receiving the response from the abort endpoint

### CODE REVIEW

File: src/main/resources/standardTestCases/restart.restart_test_without_recover.test [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ignore: true
-name: cromwell_restart_without_recover
+name: restart_tes_without_recover
```
Comment (kshakir) [inline src/main/resources/standardTestCases/restart.restart_test_without_recover.test:2]: "tes" -> "test"

File: src/main/scala/centaur/test/Test.scala [hunk 2]
```diff
@@ -142,7 +182,66 @@ object Operations {
...
+      override def run: Try[Unit] = sendReceiveFutureCompletion(() => Future { doPerform() })
+    }
+  }
+
+  def validatePAPIAborted(jobId: String, workflow: SubmittedWorkflow): Test[Unit] = {
+    import scala.concurrent.duration._
+
+    def eventually(startTime: OffsetDateTime, timeout: FiniteDuration)(f: => Try[Unit]): Try[Unit] = {
+      f match {
+        case Failure(_) if OffsetDateTime.now().isBefore(startTime.plusSeconds(timeout.toSeconds)) =>
```
Comment (kshakir) [inline src/main/scala/centaur/test/Test.scala:194]: ToL: `now` instead of `now()`?

File: src/main/scala/centaur/test/workflow/Workflow.scala [hunk 3]
```diff
@@ -37,25 +39,31 @@ object Workflow {
       case Failure(_) => invalidNel(s"Invalid test config: $path")
     }
   }
-
+
```
Comment (kshakir) [inline src/main/scala/centaur/test/workflow/Workflow.scala]: ToL: Remove the spaces

File: src/test/scala/centaur/AbstractCentaurTestCaseSpec.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import java.nio.file.Path
+
+import cats.data.Validated.{Invalid, Valid}
+import cats.instances.list._
+import cats.syntax.traverse._
+import centaur.test.standard.CentaurTestCase
+import lenthall.validation.ErrorOr.ErrorOr
+import org.scalatest.{DoNotDiscover, FlatSpec, Matchers, Tag}
+
+@DoNotDiscover
```
Comment (kshakir) [inline src/test/scala/centaur/AbstractCentaurTestCaseSpec.scala:12]: Cool!

File: src/test/scala/centaur/CentaurTestSuite.scala [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+/**
+  * The main centaur test suites, runs sub suites in parallel, but allows better control over the way each nested suite runs.
+  */
+class CentaurTestSuite extends Suites(new RestartTestCaseSpec(), new StandardTestCaseSpec()) with ParallelTestExecution with BeforeAndAfterAll {
+
+  override def beforeAll() = {
+    sys.addShutdownHook { CromwellManager.stopCromwell() }
+  }
+
+  override def afterAll() = CromwellManager.stopCromwell()
```
Comment (kshakir) [inline src/test/scala/centaur/CentaurTestSuite.scala]: Very ToL: I've always been a fan of removing the shutdown hook after it runs manually. It's a bit more code though so I understand if that extra work is skipped.

**Review by kshakir — APPROVED** [review]

File: src/main/resources/standardTestCases/restart.restart_test_without_recover.test (hunk 1 above)
Comment (Horneth) [reply src/main/resources/standardTestCases/restart.restart_test_without_recover.test:2]: it was actually meant for `TES` as TaskExecutionSchemas :p

File: src/test/scala/centaur/CentaurTestSuite.scala (hunk 5 above)
Comment (Horneth) [reply src/test/scala/centaur/CentaurTestSuite.scala]: Good idea I can add that

File: src/main/scala/centaur/test/Test.scala (hunk 2 above)
Comment (cjllanwarne) [reply src/main/scala/centaur/test/Test.scala:194]: I think `now()` is right here because it's not a pure function

Comment (mcovarr) [reply src/main/scala/centaur/test/Test.scala:194]: ooh right. it's not side effecting but it's sure as heck not referentially transparent. 😄

File: src/main/scala/centaur/test/Test.scala [hunk 6]
```diff
@@ -223,6 +322,11 @@ object Operations {
 
         for {
           actualMetadata <- CentaurCromwellClient.metadata(workflow)
+          _ = if (workflowSpec.notInMetadata.nonEmpty) {
+            // Check that none of the "notInMetadata" keys are in the actual metadata
+            val absentMdDif = workflowSpec.notInMetadata.toSet.diff(actualMetadata.value.keySet)
+            if (absentMdDif.isEmpty) throw new Exception(s"Found unwanted keys in metadata: ${absentMdDif.mkString(", ")}")
+          }
```
Comment (mcovarr) [inline src/main/scala/centaur/test/Test.scala]: that might do better as a function 🙂

File: src/test/scala/centaur/AbstractCentaurTestCaseSpec.scala (hunk 4 above)
Comment (mcovarr) [reply src/test/scala/centaur/AbstractCentaurTestCaseSpec.scala:12]: 😮

File: src/test/scala/centaur/AbstractCentaurTestCaseSpec.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    // Make tags, but enforce lowercase:
+    val tags = (testCase.testOptions.tags :+ testCase.workflow.testName :+ testCase.testFormat.name) map { x => Tag(x.toLowerCase) }
+    val isIgnored = testCase.isIgnored(cromwellBackends)
+
+    runOrDont(nameTest, tags, isIgnored, runTest)
+  }
+
+  private def runOrDont(testName: String, tags: List[Tag], ignore: Boolean, runTest: => Any): Unit = {
+
+    val itShould: ItVerbString = it should testName
```
Comment (mcovarr) [inline src/test/scala/centaur/AbstractCentaurTestCaseSpec.scala:44]: 😄

File: src/test/scala/centaur/RestartTestCaseSpec.scala [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package centaur
+
+import centaur.test.standard.CentaurTestFormat.RestartFormat
+import org.scalatest.{DoNotDiscover, Matchers}
+
+/**
+  * All test cases that trigger a Cromwell restart. Note that this suite does not mix in ParallelTestExecution
+  * such that the restarting tests execute sequentially to avoid a mayhem of Cromwell restarts
```
Comment (mcovarr) [inline src/test/scala/centaur/RestartTestCaseSpec.scala:8]: that sounds like the voice of experience 😄

**Review by mcovarr — APPROVED** [review]
very cool!

File: src/test/scala/centaur/AbstractCentaurTestCaseSpec.scala [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    def runTest = testCase.testFunction.run.get
+
+    // Make tags, but enforce lowercase:
+    val tags = (testCase.testOptions.tags :+ testCase.workflow.testName :+ testCase.testFormat.name) map { x => Tag(x.toLowerCase) }
+    val isIgnored = testCase.isIgnored(cromwellBackends)
+
+    runOrDont(nameTest, tags, isIgnored, runTest)
+  }
+
+  private def runOrDont(testName: String, tags: List[Tag], ignore: Boolean, runTest: => Any): Unit = {
```
Comment (cjllanwarne) [inline src/test/scala/centaur/AbstractCentaurTestCaseSpec.scala:42]: "There is no try"

**Pull request merged**